llama2.cu.cu

```diff
@@ -0,0 +1,726 @@
+    q[i + 1] = q0 * fci + q1 * fcr;
+    k[i] = k0 * fcr - k1 * fci;
+    k[i + 1] = k0 * fci + k1 * fcr;
+}
+
+void RoPERotation(half *q, half *k, half *f_real, half *f_imag, int num_heads, int head_size) {
+    RoPERotation_kernel <<<num_heads, head_size / 2 >>> (q, k, f_real, f_imag, num_heads, head_size);
+}
+
+__device__ void softmax_gpu(float* __restrict__ x, int size) {
+    using BlockReduce = cub::BlockReduce<float, 1024>;
+    __shared__ typename BlockReduce::TempStorage temp;
+    __shared__ float shared_val;
+
+    int tid = threadIdx.x;
+    int step = blockDim.x;
+
+    // find max value (for numerical stability)
+    float max_val = tid < size ? x[tid] : 0;
+    for (int i = tid + step; i < size; i += step)
+        if (x[i] > max_val)
+            max_val = x[i];
+
+    max_val = BlockReduce(temp).Reduce(max_val, cub::Max());
+    if (threadIdx.x == 0)
+        shared_val = max_val;
+    __syncthreads();
+    max_val = shared_val;
+
+    // exp and sum
+    float sum = 0.0f;
+    for (int i = tid; i < size; i += step) {
+        x[i] = expf(x[i] - max_val);
+        sum += x[i];
+    }
+
+    sum = BlockReduce(temp).Sum(sum);
+    if (threadIdx.x == 0)
+        shared_val = sum;
+    __syncthreads();
+    sum = shared_val;
+
+    // normalize
+    for (int i = tid; i < size; i += step)
+        x[i] /= sum;
+}
+
+// Each block processes a single head
+// Poor parallelism and even poorer memory access pattern.
+// Ankan - TODO: optimize this.
+#define MAX_SEQ_LEN 8192
+__global__ void MultiHeadAttention_kernel(half * __restrict__ output, const half* __restrict__ sq, 
+    const half* __restrict__ key_cache, const half* __restrict__ value_cache,
+    int num_heads, int head_size, int loff, int seq_len, int dim) {
+    int h = blockIdx.x;
+
+    // get the query vector for this head
+    const half* q = sq + h * head_size;
+    // attention scores for this head
+    __shared__ float att[MAX_SEQ_LEN];
+
+    // iterate over all timesteps, including the current one
+    for (int t = threadIdx.x; t < seq_len; t+= blockDim.x) {
+        // get the key vector for this head and at this timestep
+        const half* k = key_cache + loff + t * dim + h * head_size;
+        // calculate the attention score as the dot product of q and k
+        float score = 0.0f;
+        for (int i = 0; i < head_size; i++)
+            score += (float)q[i] * (float)k[i];
+        score /= sqrtf(head_size);
+        // save the score to the attention buffer
+        att[t] = score;
+    }
+    __syncthreads();
+
+    // softmax the scores to get attention weights
+    softmax_gpu(att, seq_len);
+    __syncthreads();
+
+    // weighted sum of the values, store back into xb
+    for (int i = threadIdx.x; i < head_size; i += blockDim.x) {
+        float val = 0.0f;
+        for (int t = 0; t < seq_len; t++)
+            val += att[t] * (float)value_cache[loff + t * dim + h * head_size + i];
+        output[h * head_size + i] = (half) val;
```

Comment:
the logic for this part was updated for better performance, and even readability. please check the `run.c`

Reply:
maybe this?

```c
            // weighted sum of the values, store back into xb
            half* xb = s->xb + h * head_size;
            cudaMemset(xb, 0, head_size * sizeof(half));
            for (int t = threadIdx.x; t < seq_len; t+= blockDim.x) {
                // get the value vector for this head and at this timestep
                half* v = s->value_cache + loff + t * dim + h * head_size;
                // get the attention weight for this timestep
                half a = att[t];
                // accumulate the weighted value into xb
                for (int i = 0; i < head_size; i++) {
                    xb[i] += a * v[i];
                }
            }
```

Reply:
No, the above is wrong because it needs to convert half to float, compute, and then store back as half.

It must also use `output` instead of `s->xb`

Reply:
maybe this:

```c
            // weighted sum of the values, store back into xb
            half* xb = output + h * head_size;
            cudaMemset(xb, 0, head_size * sizeof(half));
            for (int t = threadIdx.x; t < seq_len; t+= blockDim.x) {
                // get the value vector for this head and at this timestep
                half* v = s->value_cache + loff + t * dim + h * head_size;
                // get the attention weight for this timestep
                float a = att[t];
                // accumulate the weighted value into the output
                for (int i = 0; i < head_size; i++) {
                    xb[i] += (half)(a * (float)v[i]);
                }
            }
```